blog/2020/05/rr.md

```diff
@@ -0,0 +1,452 @@
+    could potentially cause a difference in execution. You really want to figure
+    out what state is relevant, since `2^2^large` is probably too many bits to send
+    somebody.
+
+2) Any input or other asynchronous events. This can mean user input, or
+    data from the network (via the NICs) or other devices. It also means things like
+    the timing of interrupts.
+
+3) Execution ordering and data races in multi-threaded executions.
+
+However, in theory if a tool was able to capture 100% of the relevant state
+from these categories, it could repeated generate exactly the same memory image.
+This is not a novel idea, but the devil is in the detail. As already mentioned,
+just capturing everything is probably too large, if at all possible (e.g. how
+do you capture interrupt timing - how do you even talk about time in this setting?).
+
+In other words, for some input state $\mathcal{S_{i}}$,
+the output state $\mathcal{S_{i+1}}$ is determined by some pure, deterministic
+function $\mathcal{I}_{ip(\mathcal{S}_i)}$, corresponding to the instruction to be
+executed at time $i$, or some asynchronous event $\mathcal{A}_i$
+
+$$ \mathcal{S}_{i+1} = \begin{cases}
+    \mathcal{A}_i(S_{i}) & \text{if an asynchronous event occured at time $i$} \\
+    \mathcal{I}_{\text{ip}(\mathcal{S}_i)} & \text{otherwise}
+\end{cases} $$
+
+Did that help? No? oh. Well, it made me feel better about my physics degree and
+also I've been told it's not research until it has at least one equation in it,
+so there you go. Also, people worked really hard to get the equation rendering working,
+on this blog, so I better use it. Anyway, where were we?
+
+Ah yes, `rr` works by treating the kernel/userspace boundary as the determinism boundary.
+This has a lot of advantages. For one, the kernel isolates a lot of state. If a
+part of the disk state (e.g. a file) wasn't requested through the kernel, you
+can be guaranteed that it didn't affect the process state (assuming the kernel
+works correctly of course). It abstracts over hardware details and provides
+a uniform interface to resources like the network. It also isolates processes
+from another, so if there is one process of interest (e.g. julia) only the
+activity relevant to it has to be recovered.
+
+Of course, there are some drawbacks as well. In order to function `rr` has to
+have an extremely precise model of the operation of the kernel and the ways
+in which it interacts with userspace. On Linux this interface is supposedly
+stable, but few applications put as stringent a constraint on that promise as
+`rr` does (e.g. even a single bit difference can be problematic).
+
+The details here are quite involved, and I recommend reading Robert's
+[technical report](https://arxiv.org/abs/1610.02144) for a deeper overview of
+some of `rr`'s design points (though even that only scratches the surface).
+
+One final thought for this section is the origin of the term "time-traveling"
```

Comment:
Apparently the only use of this term?
```suggestion
This feature of the `rr` tool is called "time-traveling" debugging.
```

Reply:
Applied.

Reply:
I'm pretty sure the term predates rr by quite a bit.